providers/implementations/ciphers/ciphercommon.c

```diff
@@ -646,8 +646,9 @@ int ossl_cipher_generic_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 int ossl_cipher_generic_initiv(PROV_CIPHER_CTX *ctx, const unsigned char *iv,
                                size_t ivlen)
 {
-    if (ivlen != ctx->ivlen
-        || ivlen > sizeof(ctx->iv)) {
+    if (ivlen > sizeof(ctx->iv)
+        || (!ctx->pad_iv && ivlen != ctx->ivlen)
+        || (ctx->pad_iv && ivlen > ctx->ivlen)) {
```

Comment:
I'd probably write the condition as:

```suggestion
        || (!ctx->pad_iv && ivlen != ctx->ivlen)
        || ivlen > ctx->ivlen) {
```

But the existing is OK too.